cluster-autoscaler/proposals/gpu_optimization.md

```diff
@@ -0,0 +1,88 @@
+
+The problem is GPU node becomes ready first but at that time, there's no allocable GPU resources. The root reason could be  
+
+* Device Plugin is not ready.
+* Device Plugin is ready but has not sent list of devices to kubelet. 
+* kubelet takes time to advertise resources to API server so that node spec is not up to date.
+
+Pod thhat requests GPU can not be scheduled during this period. In order to address this issue, we either need to work with Nvidia folks on device plugin to change behavior on upstream, or make CA mark nodes with GPU label and unallocated GPU resource as NonReady.
+
+
+#### GPU Resource is not given consideration in scaling down
+CA filter candidates with low node cpu and memory utilization rate in every loop. 
+
+* If utilization rate is high, even all GPUs are idle, scale-down will not be triggered.  
+Consider use case for machine learning inference or some training case that could tolerant failures, we actually like to give a try to see if all workloads on that particular node can be moved to other nodes. Then user can scale down this GPU node and efficiently reduce their cost.
+
+* If utilization rate is low, GPUs are in use, scale-down will be triggered.  
+If there's a distributed training task on that node, killing task will lead to entire training job failing. In this case, this node can not be a scale down candicate.
+
+```
+// CalculateUtilization calculates utilization of a node, defined as maximum of (cpu, memory) utilization.
+// Per resource utilization is the sum of requests for it divided by allocatable. It also returns the individual
+// cpu and memory utilization.
+func CalculateUtilization(node *apiv1.Node, nodeInfo *schedulercache.NodeInfo, skipDaemonSetPods, skipMirrorPods bool) (utilInfo UtilizationInfo, err error) {
+    cpu, err := calculateUtilizationOfResource(node, nodeInfo, apiv1.ResourceCPU, skipDaemonSetPods, skipMirrorPods)
+    if err != nil {
+        return UtilizationInfo{}, err
+    }
+    mem, err := calculateUtilizationOfResource(node, nodeInfo, apiv1.ResourceMemory, skipDaemonSetPods, skipMirrorPods)
+    if err != nil {
+        return UtilizationInfo{}, err
+    }
+    return UtilizationInfo{CpuUtil: cpu, MemUtil: mem, Utilization: math.Max(cpu, mem)}, nil
+}
+
+```
+> node utilization calculation logic
+
+
+### Proposed Solution
+1. Either move `utils/gpu` logic to cloud provider or have a new option passed from commandline to indicate gpu node label cloud provider like to use
+2. Scale down case is kind of tricky because training and serving seems like different use case and have confliction. Fit GPU resource into utilization formula doesn't solve this issue. Instead, I think it's better to have a flag to indicate if gpu nodes can be scaled down or not.
```

Comment:
Why using GPU in resource utilization formula doesn't solve the issue? The problem I'm aware of is high CPU or memory utilization preventing scale-down of nodes with GPU. This can be solved with changing utilization calculation to only care about GPU as described in https://github.com/kubernetes/autoscaler/issues/1367#issuecomment-449358176. 

Note that utilization check is just one condition for scale-down (and one that we should ideally get rid of at some point anyway). Having low utilization doesn't mean a node will actually be deleted. We have plenty of mechanisms for preventing scale-down if you don't want particular node removed (scale-down-disabled annotation) or specific pod restarted (safe-to-evict: false annotation, PDB to name just a few).
Unless there is a use-case that can't be addressed with existing mechanisms I'd rather we improve documentation and recommend an existing solution, rather than add another flag.

Reply:
For GPU node, in training case, GPU utilization rate is not helpful to make decision.
For example, given a 8 GPU machine, there're 2 tasks and one take 6 GPUs and the other takes 2 GPUs. Once first task finishes and GPU resource release. The GPU utilization rate is 25%. But it can not be killed because failure of this task will result in a failure of entire training job. 

That's why I think ignore CPU and memory utilization formula can not resolve this issue. 
GPU utilization rate will help on model inference case. Most of workloads are stateless and can be easily migrated to other nodes to reduce cost.

Reply:
I look at the comment and it's reasonable to create PDB and annotate node. The speciality is most of batch/ML workloads are short term lived. It's kind of tedious to do that if jobs created frequently.

Reply:
I think we can make training case to more generic, like a flag to indicate if we can remove a node immediately (kill in-flight job) or wait until workloads done.  Looks like this is not supported but important for batch jobs.

Reply:
"Note that utilization check is just one condition for scale-down (and one that we should ideally get rid of at some point anyway)." Do you mean in the future CA has plan to add some other triggers? I think entire cluster node utilization might be a reasonable one. That tradeoff here is scale up speed vs cost. Do you think this feature can be added?

Reply:
Utilization is not a trigger, it's only the first step in scale-down decision, a precondition. The other part is 'drain simulation' that calculates if all pods are safe to move (if they fit elsewhere in the cluster, if they're owned by controller that will restart them, if PDB allows moving them, all the stuff mentioned in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node).

The drain part is the important one, the utilization part is there to limit the number of evictions (don't restart the pods if the gain is relatively small) and for performance reasons (avoid expensive drain simulation if it's unlikely the node can be deleted).  

Regarding your use-case: it seems like putting "cluster-autoscaler.kubernetes.io/safe-to-evict": "false" on your pods is a simple solution.

Reply:
Agree. Use label per job for fine grained control makes sense. At the beginning, I thought covering both cases using cluster level tag which may makes it a little bit confused and contradictory somehow.